### PR TITLE
Buddha mode + Hell

### PR DESCRIPTION
**Pull request by SECBATON-GRIFFON**
[content][administration]

## What this does
- [x] creates "buddha mode", similar to god mode but the affected player still takes all damage, feels all pain, gets all status effects and etc., but does not become unconscious or die. name i think is from the "buddha" console command cheat on a lot of valve games
- [x] also creates a custom z2 dungeon called "hell" and the option for admins to send any player ghost there, creating a copy of their body they get transferred to with buddha mode on teleported into a large lake of fire. the dungeon only gets created when the first player is sent there.
- [x] populates hell with the souls of previously permabanned players, based on their appearance savefiles

## Why it's good
a unique kind of admin punishment

## Changelog
:cl:
 * rscadd: Admins can now toggle "buddha mode" on players, preventing them from ever becoming unconscious or dead while still allowing them to take damage or feel pain/ailments.
 * rscadd: Admins can now send particularly guilty dead souls to the realm of H-E-double hockey sticks for their crimes in life.

### CODE REVIEW

**Comment by jwhitak** [comment]
So what exactly happens when they get gibbed? Holy...

**Comment by SECBATON-GRIFFON** [comment]
> So what exactly happens when they get gibbed? Holy...

theres a lot of edge cases like this i need to account for (hence draft at time of this comment) so maybe ill just prevent gibbing?
need to think about how this can affect cluwnes too

**Comment by gurfan** [comment]
~~What's the point?~~

holy shit

**Comment by SECBATON-GRIFFON** [comment]
> What's the point?

now you see, hopefully

**Comment by ExpulsionSnooper** [comment]
If you're going to add hell, remove the option to ghost for people sent there, PLEASE.

**Comment by smalIboss** [comment]
thank you

**Comment by SECBATON-GRIFFON** [comment]
> If you're going to add hell, remove the option to ghost for people sent there, PLEASE.

say no more fam

**Comment by Eneocho** [comment]
Can you make it so a hell-punished player can go through "purgatory" if given the chance (by an admeme) and revive as a randomname randombody?

>populate hell with the souls of previously permabanned players, based on their appearance savefiles

Unholy, demonic, abandoned by god levels of based.
I kneel.

**Comment by tinysaturn** [comment]
Requesting a frozen/ice version

**Comment by SECBATON-GRIFFON** [comment]
> Requesting a frozen/ice version

maybe for april fools

**Comment by NiftyRyanyan** [comment]
Rolling a one on the cursed dice should send you to hell.

**Comment by SECBATON-GRIFFON** [comment]
> Rolling a one on the cursed dice should send you to hell.

done, on a 1% chance. prepare for unforeseen consequences.

**Comment by ghost** [comment]
finally a reason to add ultrakill content

**Comment by SECBATON-GRIFFON** [comment]
> So what exactly happens when they get gibbed? Holy...

gibbing has been disabled while on it, instead it causes damage. heads are also immune from being gibbed, while other limbs aren't

**Comment by Eneocho** [comment]
Feels kinda half assed to have gibbing disabled on the torso and head but not the arms and legs. Disable gibbing on the arms and legs as well for extra pain. More limbs to suffer with.

**Comment by ManopolisJones** [comment]
I feel like you're gonna need this to really hammer home the point

IF THIS BANNER IS FLASHING

**Comment by SECBATON-GRIFFON** [comment]
if anyone knows how to make some mockup player pref and ban database stuff to test loading banned players in it'd be much help for this to continue, otherwise if i get no answer in a week i'll disable it and undraft

**Comment by jwhitak** [comment]
It is one week, you have no answer, you can now undraft.

**Comment by Eneocho** [comment]
I think the test server has its own ban list, you could testmerge and use the already implemented list.

File: code/modules/randomMaps/vaults.dm [hunk 1]
```diff
@@ -102,6 +102,209 @@
 
 	message_admins("<span class='info'>Loaded space hobo shack [result ? "" : "un"]successfully.</span>")
 
+/proc/create_hell()
```
Comment (Kurfursten) [inline code/modules/randomMaps/vaults.dm]: it seems insane to create hell every shift when we rarely even use centcomm jail

just make this a fun verb of its own

Comment (Eneocho) [reply code/modules/randomMaps/vaults.dm]: The difference there is that the centcomm jail is basically a no interaction totally bland cuck cube. Hell is on fire, has colour, you can scream your lungs out and, if he eventually implements it, the spirits and tortured bodies of shitters of olde.

Comment (gurfan) [reply code/modules/randomMaps/vaults.dm]: > The difference there is that the centcomm jail is basically a no interaction totally bland cuck cube. Hell is on fire, has colour, you can scream your lungs out and, if he eventually implements it, the spirits and tortured bodies of shitters of olde.

That doesn't mean it needs to be created automatically every shift.

Comment (Exxion) [reply code/modules/randomMaps/vaults.dm]: Judging by the OP, I'm guessing its automatic creation was left in by accident

Comment (Eneocho) [reply code/modules/randomMaps/vaults.dm]: >That doesn't mean it needs to be created automatically every shift.

True. Perhaps the centcomm jail should be replaced with hell on each map?

Comment (Exxion) [reply code/modules/randomMaps/vaults.dm]: It's already coded to generate hell if required when the first person is sent to it. The automatic generation was probably left in by mistake.

Comment (SECBATON-GRIFFON) [reply code/modules/randomMaps/vaults.dm]: nah not a mistake, it freezes the game for 4 seconds at the lowest (was previously way longer until i did the fast turf init thing) so less people notice it at roundstart load

Comment (Exxion) [reply code/modules/randomMaps/vaults.dm]: Then perhaps you should remove the text from the OP stating that it's only generated when a person is first sent there. Also, pregenerating a whole Z-level to be used once a year or so is pretty unacceptable.

Comment (Kurfursten) [reply code/modules/randomMaps/vaults.dm]: The concept is great, but adding +4 seconds to load time every shift forever when it won't seriously be used is not good. If you really want to, you can map something onto z-2 manually where there is plenty of space.

Comment (Eneocho) [reply code/modules/randomMaps/vaults.dm]: >you can map something onto z-2 manually where there is plenty of space

Careful to not map over the vault areas, some of the larger vaults, such as sokoban or the mothership lab take a LOT of space, you don't want the mapped hell to be overwritten by the vaults (or viceversa).

Comment (Exxion) [reply code/modules/randomMaps/vaults.dm]: Or you could generate it in the Z2 vault area
I don't remember creating a new Z-level taking four seconds, though. Maybe I'm just remembering incorrectly, but I swear it was much faster. Though it is highly hardware-dependent too, of course.

EDIT: On my machine, generating a new Z-level takes a fraction of a second, causing a barely-noticeable hitch.

Comment (NigglyWiggly) [reply code/modules/randomMaps/vaults.dm]: If generating a new Z level every round is gonna be such a probblem, you could just as easily make Hell a corner of the Centcomm Z-Level or something
Smaller scale hell, same suffering
Ideally we'd be able to make it loop, just a tiny looping box of hell in the corner

Comment (SECBATON-GRIFFON) [reply code/modules/randomMaps/vaults.dm]: making it a dungeon to resolve performance issues

**Comment by SECBATON-GRIFFON** [comment]
okay so copying prefs actually is tested working using a local db, doing this more than once shouldn't lag anything too bad so i'm actually gonna go ahead with the banned player thing, but only one for now to test performance.

**Comment by Eneocho** [comment]
> but only one for now to test performance.

I feel a random sample from 1-5 would've been better to test performance, since you could check how many did spawn and how laggy did it get.

**Pull request merged**